test/index.js

```diff
@@ -1424,4 +1432,163 @@ describe('restify-client tests', function () {
             });
         });
     });
+
+    // Ensure the .before and .after functions are called for each client.
+    describe('before/after hooks', function () {
+        var _clients = {
+            HttpClient: {
+                creator: clients.createHttpClient
+            }, JsonClient: {
+                creator: clients.createJsonClient
+            }, StringClient: {
+                creator: clients.createStringClient
+            }
+        };
+        var handles = [];
+        var server = restify.createServer({});
+        var serverUrl;
+
+        function _getBody(clientType, req, data, cb) {
+            if (clientType === 'HttpClient') {
+                req.on('result', function (err, res) {
+                    var body = '';
+
+                    assert.ifError(err);
+
+                    res.setEncoding('utf8');
+                    res.on('data', function (chunk) {
+                        body += chunk;
+                    });
+
+                    res.on('end', function () {
+                        cb(body.replace(/"/g, ''));
+                    });
+                });
+            } else {
+                cb(data.replace(/"/g, ''));
+            }
+        }
+
+        before(function (callback) {
+            var url = 'http://127.0.0.1';
+
+            server.get('/testheader', sendTestHeader);
+            server.listen();
+            serverUrl = url + ':' + server.address().port;
+
+            callback();
+        });
+
+        // {Http,Json,String}Client should be able to use .before and .after
+        Object.keys(_clients).forEach(function (clientName) {
+            var _client = _clients[clientName];
+
+            it('request from ' + clientName, function (done) {
+                var afterCtx = false;
+                var afterRan = false;
+                var beforeRan = false;
+                var client;
+
+                client = _client.creator({
+                    after: function (err, req, res, ctx, cb) {
+                        assert.ifError(err);
+                        afterRan = true;
+                        // got ctx from before() function
+                        if (ctx && ctx.hello === 'world') {
+                            afterCtx = true;
+                        }
+                        cb();
+                    }, before: function (opts, cb) {
+                        beforeRan = true;
```

Comment:
can we add more assertation around opts?
To be sure that this is the complete `opts` object, not just the subset of it.